source/extensions/filters/http/oauth/filter.cc

```diff
@@ -0,0 +1,475 @@
+  user_agent_ = v[0];
+
+  sanitizeXForwardedOauthHeaders(headers);
+
+  // We should check if this is a sign out request.
+  if (path_str == config_->signoutPath()) {
+    return signOutUser(headers);
+  }
+
+  if (canSkipOAuth(headers)) {
+    // Update the path header with the query string parameters after a successful OAuth login.
+    // This is necessary if a website requests multiple resources which get redirected to the
+    // auth server. A cached login on the authentication server side will set cookies
+    // correctly but cause a race condition on future requests that have their location set
+    // to the callback path.
+    if (absl::StartsWith(path_str, config_->callbackPath())) {
+      Http::Utility::QueryParams query_parameters = Http::Utility::parseQueryString(path_str);
+
+      const auto state =
+          absl::StrReplaceAll(query_parameters.at(queryParamsState()), unescapedReplacements());
+      Http::Utility::Url state_url;
+      if (!state_url.initialize(state, false)) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      // Avoid infinite redirect storm
+      if (absl::StartsWith(state_url.pathAndQueryParams(), config_->callbackPath())) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      Http::ResponseHeaderMapPtr response_headers{
+          Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+              {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))},
+               {Http::Headers::get().Location, state}})};
+      decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+    }
+
+    // Continue on with the filter stack.
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // If a bearer token is supplied as a header or param, we ingest it here and kick off the
+  // user resolution immediately. Note this comes after HMAC validation, so technically this
+  // header is sanitized in a way, as the validation check forces the correct Bearer Cookie value.
+  access_token_ = extractAccessToken(headers, Token);
+  if (!access_token_.empty()) {
+    found_bearer_token_ = true;
+    request_headers_ = &headers;
+    oauth_client_->asyncGetIdentity(access_token_);
+
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  // If no access token and this isn't the callback URI, redirect to acquire credentials.
+  //
+  // The following conditional could be replaced with a regex pattern-match,
+  // if we're concerned about matching strictly `/_oauth`.
+  if (!absl::StartsWith(path_str, config_->callbackPath())) {
+    Http::ResponseHeaderMapPtr response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+        {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))}})};
+
+    // Construct the correct scheme. We default to https since this is a requirement for OAuth to
```

Comment:
required for this implementation in particular? required for the call to the authz server?

Reply:
It's not technically required for the redirection endpoint, but it is required for the authorize and token endpoints so it's highly likely that HTTPS is supported.

From the RFC:
```
  The redirection endpoint SHOULD require the use of TLS as described
   in Section 1.6 when the requested response type is "code" or "token",
   or when the redirection request will result in the transmission of
   sensitive credentials over an open network.  This specification does
   not mandate the use of TLS because at the time of this writing,
   requiring clients to deploy TLS is a significant hurdle for many
   client developers.  If TLS is not available, the authorization server
   SHOULD warn the resource owner about the insecure endpoint prior to
   redirection (e.g., display a message during the authorization
   request).
```

We could either make it a boolean field defaulting to true, or go as-is and wait for it to be requested.